### PR TITLE
Comments, trailing commas, flags, and optimizations

### PR DESCRIPTION
**Pull request by sdkrystian**

### CODE REVIEW

File: include/boost/json/detail/sse2.hpp [hunk 1]
```diff
@@ -1,6 +1,7 @@
 //
 // Copyright (c) 2019 Peter Dimov (pdimov at gmail dot com),
 //                    Vinnie Falco (vinnie.falco@gmail.com)
+// Copyright (c) 2020 Krystian Stasiowski (sdkrystian@gmail.com)
```
Comment (sdkrystian) [inline include/boost/json/detail/sse2.hpp:4]: Oops, forgot to delete this

File: bench/bench.cpp [hunk 2]
```diff
@@ -320,7 +320,7 @@ class boost_pool_impl : public any_impl
         parser p;
         while(repeat--)
         {
-            monotonic_resource mr;
+            monotonic_resource mr(1024 * 64);
```
Comment (vinniefalco) [inline bench/bench.cpp:323]: Does this make a difference?

Comment (vinniefalco) [reply bench/bench.cpp:323]: I think this should be in its own commit because it affects performance, and it is currently buried in a larger optimization commit.

File: bench/bench.cpp [hunk 3]
```diff
@@ -374,9 +374,9 @@ class boost_vec_impl : public any_impl
         bool on_document_begin(error_code&) { return true; }
         bool on_document_end(error_code&) { return true; }
         bool on_object_begin(error_code&) { return true; }
-        bool on_object_end(std::size_t, error_code&) { return true; }
+        bool on_object_end(error_code&) { return true; }
         bool on_array_begin(error_code&) { return true; }
-        bool on_array_end(std::size_t, error_code&) { return true; }
+        bool on_array_end(error_code&) { return true; }
```
Comment (vinniefalco) [inline bench/bench.cpp:379]: Removal of the `size_t` should be in its own commit

**Comment by sdkrystian** [comment]
@vinniefalco when this is merged just cherry-pick the "Comment and trailing comma support" commit after merging the previous two in the other PR

File: bench/bench.cpp (hunk 2 above)
Comment (sdkrystian) [reply bench/bench.cpp]: Copy

Comment (sdkrystian) [reply bench/bench.cpp]: The difference made is negligible -- I added this to make our benchmark more similar to rapidjson

**Comment by vinniefalco** [comment]
1. The variable name for parse options should be `opt` or `opt_`.

2. There are merge conflicts, please rebase

**Comment by sdkrystian** [comment]
@vinniefalco Done and done

File: include/boost/json/detail/basic_parser.hpp [hunk 4]
```diff
@@ -105,6 +107,7 @@ class basic_parser
     bool more_; // false for final buffer
     bool complete_ = false; // true on complete parse
     bool is_key_;
+    parse_options opt_ = {false, false};
```
Comment (vinniefalco) [inline include/boost/json/detail/basic_parser.hpp:110]: But... `parse_options` already has these set to false in the default ctor?

File: include/boost/json/parse_options.hpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+/** Parser Extensions.
+
+*/
+struct parse_options
+{
+    bool allow_comments = false;
+    bool allow_trailing_commas = false;
+};
+
+#endif
```
Comment (vinniefalco) [inline include/boost/json/parse_options.hpp:22]: missing newline

File: include/boost/json/parse_options.hpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/cppalliance/json
+//
+
+#ifndef BOOST_JSON_PARSE_OPTIONS_HPP
+#define BOOST_JSON_PARSE_OPTIONS_HPP
+
+/** Parser Extensions.
```
Comment (vinniefalco) [inline include/boost/json/parse_options.hpp:13]: There will eventually need to be more written here

Comment (vinniefalco) [reply include/boost/json/parse_options.hpp:13]: You could say "the default parse options allow only strict JSON."

File: include/boost/json/basic_parser.hpp [hunk 7]
```diff
@@ -45,6 +46,10 @@ namespace json {
 enum class basic_parser::state : char
 {
     doc1, doc2, doc3,
+    com1, com2, com3,
+    com4, com5, com6, 
+    com7, com8, com9, 
+    com10, com11, com12,
```
Comment (vinniefalco) [inline include/boost/json/basic_parser.hpp:52]: TRIGGERED!

File: include/boost/json/basic_parser.hpp [hunk 8]
```diff
@@ -305,8 +456,22 @@ parse_document(
...
+        result r;
+        if (opt_.allow_comments)
+            if (opt_.allow_trailing_commas)
+                r = parse_value<StackEmpty, true, true>(h, cs);
+            else
+                r = parse_value<StackEmpty, true, false>(h, cs);
+        else if (opt_.allow_trailing_commas)
+            r = parse_value<StackEmpty, false, true>(h, cs);
+        else
+            r = parse_value<StackEmpty, false, false>(h, cs);
```
Comment (vinniefalco) [inline include/boost/json/basic_parser.hpp:474]: I have to insist on braces, e.g.
```
if (opt_.allow_comments)
{
  if (opt_.allow_trailing_commas)
...
```

File: include/boost/json/detail/basic_parser.hpp (hunk 4 above)
Comment (sdkrystian) [reply include/boost/json/detail/basic_parser.hpp:110]: This is a bit of a dilemma -- while `parser` does this, the default constructor for `basic_parser` will not. We can change it though

**Comment by vinniefalco** [comment]
Still failing

File: include/boost/json/parse_options.hpp (hunk 6 above)
Comment (sdkrystian) [reply include/boost/json/parse_options.hpp:13]: I've added a bit.
